osu.Game/Tests/Visual/OsuTestScene.cs

```diff
@@ -124,6 +139,14 @@ public virtual void RecycleLocalStorage()
         [Resolved]
         protected AudioManager Audio { get; private set; }
 
+        /// <summary>
+        /// Creates the ruleset to be used for this test scene.
+        /// </summary>
+        /// <remarks>
+        /// When testing against ruleset-specific components, this method must be overriden to their ruleset.
+        /// </remarks>
+        protected virtual Ruleset CreateRuleset() => null;
```

Comment:
Maybe mark as
```suggestion
        [CanBeNull]
        protected virtual Ruleset CreateRuleset() => null;
```
which reveals a potentially unsafe access at `PlayerTestScene`, which in turn I would resolve with
```csharp
        [NotNull]
        protected abstract Ruleset CreatePlayerRuleset();

        protected sealed override Ruleset CreateRuleset() => CreatePlayerRuleset();
```

Reply:
After rethinking this for a bit, seems like going with your suggestion is the best way to have a good access to `Ruleset.Value`.